openedx/openedx-platform
# feat: Add support for pinning and closing threads, adds a preview and a delete check field [BD-38] [TNL-8719]

## Description

- Adds support for pinning and closing threads to the current REST APIs for discussions.
- Adds a preview_body field to the thread response that contains a truncated version of the raw body with tags removed so it can be shown in previews.
- Adds a field called "can_delete" to threads and comments that signals if the current user is able to delete that resource.

This first allows the [new discussions MFE](https://github.com/edx/frontend-app-discussions) to handle pinning and closing threads.

The second allows a shorter preview text to be displayed for each post/thread by pre-rendering this in the API. 

The final point makes it simpler to determine if a post is deletable in the frontend.  Other actions can be represented in `editable_fields` however, delete cannot. 

## Supporting information

- [TNL-8719](https://openedx.atlassian.net/browse/TNL-8719)

## Testing instructions

- Create some test data for the forums
- Use the rest API at `/api/discussion/v1/` to update an existing thread toggling `pinned` and `closed` both should work if the current user has access
- Check that the `preview_body` contains a useful preview of the content
- Check that `can_delete` accurately represent the conditions in which a comment or thread can be deleted, which is that the user is either privileged or the author of a post.

## Deadline

"None"
